core/unit_test/tools/TestWithoutInitializing.cpp

```diff
@@ -47,25 +47,43 @@
 
 #include "include/ToolTestingUtilities.hpp"
 
+template <typename InputView, typename OutputView>
+void check_host_mirror(const InputView&, const OutputView&) {
+  using InputSpace  = typename InputView::memory_space;
+  using OutputSpace = typename OutputView::memory_space;
+  if constexpr (Kokkos::SpaceAccessibility<Kokkos::HostSpace,
+                                           InputSpace>::accessible)
+    static_assert(std::is_same_v<OutputSpace, InputSpace>);
+  else
+    static_assert(std::is_same_v<OutputSpace, Kokkos::HostSpace>);
+}
```

Comment:
I did some reading on this recently that leads me to think an unsatisfied `static_assert` in the discarded branch of an `if constexpr` makes the program 'ill formed, no diagnostic required'. I'm not entirely sure if I was reading the exact conditions involved correctly though.

Conversely, the non-C++17 version of this that doesn't use `if constexpr` would raise no such concerns. We'd probably want to make sure that the whole team understands them for future use cases, though.

With that in mind, could you stick each of the expressions in `constexpr bool` variables, and then just use them in a single `static_assert`?


Something like
```
(is_accessible && is_input) || (!is_accessible && is_host)
```

Reply:
Now that I've written that out, I think a ternary might be clearer

Reply:
> I did some reading on this recently that leads me to think an unsatisfied static_assert in the discarded branch of an if constexpr makes the program 'ill formed, no diagnostic required'. I'm not entirely sure if I was reading the exact conditions involved correctly though.

Would you mind finding a reference to that? https://stackoverflow.com/a/62469617 cites the standard to say
> The validity of a template may be checked prior to any instantiation. [...] The program is ill-formed, no diagnostic required, if:
>
>  - no valid specialization can be generated for a template or a substatement of a constexpr if statement within a template >and the template is not instantiated, or...
> [...]

but here there are cases where both branches are valid.